### ISSUE TITLE
Spurious TypeVar appears only once in a generic function signature with `typing.Callable` and `typing.overload`

### ISSUE THREAD

**Issue by sauyon** [microsoft/pyright#3397]
Pylance reports an invalid TypeVar use for the following snippet:

```python
import typing

T = typing.TypeVar("T", bound=typing.Callable[..., typing.Any])


class A:
    @typing.overload
    def foo(self, a: T) -> T:
        pass

    @typing.overload
    def foo(self, a: int) -> typing.Callable[[T], T]:
        pass

    def foo(self, a: int | T) -> typing.Callable[[T], T] | T:
        print(a)
        raise(NotImplementedError)
```

```
[{
	"resource": "/s:/Documents/Development/BentoML/test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"code": {
		"value": "reportInvalidTypeVarUse",
		"target": {
			"$mid": 1,
			"external": "https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules",
			"path": "/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "diagnostic-severity-rules"
		}
	},
	"severity": 8,
	"message": "TypeVar \"T\" appears only once in generic function signature",
	"source": "Pylance",
	"startLineNumber": 15,
	"startColumn": 28,
	"endLineNumber": 15,
	"endColumn": 29
}]
```

I've also tried making a new type variable to use with the second overload, but that didn't seem to solve the issue.

The issue itself seems weirdly non-deterministic; while doing the above and playing around with the body of the actual `foo` method I've occasionally seen the error disappear.

**Expected behavior**
No error.

**VS Code extension or command-line**
VS Code Pylance v2022.4.3

**Comment by erictraut** [comment]
I'm not able to repro the problem you're reporting. Please verify that when you copy and paste the code sample above into a new document within VS Code that you see an error.

The error message you mentioned above doesn't make sense given the code snippet, so I suspect that you maybe simplified the code to the point where the error went away.

**Comment by erictraut** [comment]
Oh, I was just able to repro the problem by forcing a different evaluation order. I can do this in pyright by hovering over the return type annotation in the implementation while modifying the document. The order of evaluation shouldn't affect the results, so this is definitely a bug. I'll investigate further.

**Comment by erictraut** [comment]
This was a long-standing bug. I'm surprised it hasn't been reported previously. It will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.242, which I just published. It will also be included in the next release of pylance.
